core/src/main/scala/org/http4s/Uri.scala

```diff
@@ -29,10 +29,15 @@ final case class Uri(
   fragment: Option[Fragment] = None)
   extends QueryOps with Renderable
 {
+  import Uri._
+
   def withPath(path: Path): Uri = copy(path = path)
 
-  def /(newFragment: Path): Uri = {
-    val newPath = if (path.isEmpty || path.last != '/') path + "/" + newFragment else path + newFragment
+  def /(newSegment: Path)(implicit charset: Charset = Charset.`UTF-8`): Uri = {
```

Comment:
Is there any case in which it makes sense to have different segments encoded with different charsets? I suspect not, which makes me wonder if the `charset` should be part of the class as opposed to an argument to this method. But then it also seems like you shouldn't need a charset for an empty path, and I guess it's probably safe to assume that people will use the same charset throughout construction of their URL.

I think there's another option which is more breaking but possibly more elegant. What if `path` were something like `List[Segment]`? Then you could have a couple options. You could just have `type Segment = String` and require a charset when you try to render the path as a string. Or you could have something like:

```scala
final class Segment private(val encoded: String) extends AnyVal

object Segment {
  /** Encodes the string with the provided charset */
  def apply(s: String)(implicit charset: Charset = Charset.`UTF-8`): Segment = ...

  /** Create a segment without encoding it. This is useful if your segment is already encoded or you know that it doesn't need to be (because it is an `Int`, `Long`, `Boolean`, etc. */
  def noEncoding(s: String): Segment = new Segment(s)
}
```

You could even go a step further and create a type class for types that can be encoded for path segments and then you could have something like:

```scala
def /[A](a: A)(implicit pathEncoder: PathEncoder[A]): Uri = ...
```

And then people could pass in a `String`, a `Long`, etc in path segments and we (the type class instances) could be smart enough to know that if it's a `String` we should encode it, but if it's a `Long` we can just call `toString`.

This is just a stream of consciousness. Feel free to completely disregard all of it.